marine.c

```diff
@@ -966,6 +977,374 @@ WS_DLL_PUBLIC void marine_free(marine_result *ptr) {
     }
 }
 
+static void marine_packet_field_join_children_with_duplicate_names(marine_packet_field *field);
+static void marine_packet_field_free(marine_packet_field *field);
+
+static marine_packet_field *
+marine_packet_field_init(marine_packet_field *node, char *name) {
+    node->name = name;
+    node->children = NULL;
+    node->value.type = MARINE_VT_NONE,
+    node->value.len = 0;
+    return node;
+}
+
+static void
+marine_packet_field_add_child(marine_packet_field *parent, marine_packet_field *child) {
+    if (parent->children == NULL) {
+        parent->children = g_array_new(TRUE, TRUE, sizeof(marine_packet_field));
+    }
+    g_array_append_val(parent->children, *child);
+}
+
+static marine_packet_field_value
+marine_packet_field_load_value(field_info *finfo) {
+    ftenum_t ftype = finfo->value.ftype->ftype;
+    marine_packet_field_value value;
+    if (ftype == FT_FLOAT || ftype == FT_DOUBLE) {
+        value.type = MARINE_VT_FLOAT;
+        value.len = finfo->value.ftype->wire_size;
+        value.float_value = fvalue_get_floating(&finfo->value);
+    } else if (IS_FT_INT(ftype)) {
+        value.type = MARINE_VT_INT;
+        value.len = finfo->value.ftype->wire_size;
+        if (IS_FT_INT32(ftype)) {
+            value.int_value = fvalue_get_sinteger(&finfo->value);
+        }
+        if (IS_FT_INT64(ftype)) {
+            value.int_value = fvalue_get_sinteger64(&finfo->value);
```

Comment:
Isn't this losing precision on 32bit machines? (same for the `uint` handling)

Reply:
Right, I'll make `int_value` and `uint_value` to be `long`.